packages/react-components/react-datepicker-compat/src/components/DatePicker/useDatePicker.tsx

```diff
@@ -358,24 +362,31 @@ export const useDatePicker_unstable = (props: DatePickerProps, ref: React.Ref<HT
       contentAfter: <CalendarMonthRegular onClick={onIconClick as unknown as React.MouseEventHandler<SVGElement>} />,
       readOnly: !allowTextInput,
       role: 'combobox',
-      root: {
-        'aria-owns': open ? popupSurfaceId : undefined,
-        ref: useMergedRefs(triggerWrapperRef, ref),
-      },
-      input: {
-        ref: rootRef,
-      },
     },
   });
 
-  rootShorthand.onChange = mergeCallbacks(rootShorthand.onChange, onInputChange);
-  rootShorthand.onBlur = mergeCallbacks(rootShorthand.onBlur, onInputBlur);
-  rootShorthand.onKeyDown = mergeCallbacks(rootShorthand.onKeyDown, onInputKeyDown);
-  rootShorthand.onFocus = mergeCallbacks(rootShorthand.onFocus, onInputFocus);
-  rootShorthand.onClick = mergeCallbacks(rootShorthand.onClick, onInputClick);
+  const rootShorthand = resolveShorthand(props.root, {
+    required: true,
+    defaultProps: {
+      'aria-owns': open ? popupSurfaceId : undefined,
+      ref: triggerWrapperRef,
+    },
+  });
+  rootShorthand.ref = useMergedRefs(rootShorthand.ref, triggerWrapperRef);
+  root.root = rootShorthand;
+
+  const inputShorthand = resolveShorthand(props.input, { required: true });
+  inputShorthand.ref = useMergedRefs(ref, rootRef, inputShorthand.ref);
```

Comment:
Is there a reason `inputShorthand.ref` is last here but `rootShorthand.ref` is first above? `useMergedRef` invokes refs in the order in which they are provided to the hook and I'd generally expect the user-provided ref to be invoked before anything else.